crates/storage/nippy-jar/src/cursor.rs

```diff
@@ -0,0 +1,146 @@
+use crate::{compression::Compression, Filter, NippyJar, NippyJarError, PerfectHashingFunction};
+use memmap2::Mmap;
+use std::{clone::Clone, fs::File, sync::Mutex};
+use sucds::int_vectors::Access;
+use zstd::bulk::Decompressor;
+
+/// Simple cursor implementation to retrieve data from [`NippyJar`].
+pub struct NippyJarCursor<'a> {
+    /// [`NippyJar`] which holds most of the required configuration to read from the file.
+    jar: &'a NippyJar,
+    /// Optional dictionary decompressors.
+    zstd_decompressors: Option<Mutex<Vec<Decompressor<'a>>>>,
+    /// Data file.
+    #[allow(unused)]
+    file_handle: File,
+    /// Data file.
+    mmap_handle: Mmap,
+    /// Temporary buffer to unload data to (if necessary), without reallocating memory on each
+    /// retrieval.
+    tmp_buf: Vec<u8>,
+    /// Cursor row position.
+    row: u64,
+    /// Cursor column position.
+    col: u64,
+}
+
+impl<'a> std::fmt::Debug for NippyJarCursor<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NippyJarCursor {{ config: {:?} }}", self.jar)
+    }
+}
+
+impl<'a> NippyJarCursor<'a> {
+    pub fn new(
+        config: &'a NippyJar,
+        zstd_decompressors: Option<Mutex<Vec<Decompressor<'a>>>>,
+    ) -> Result<Self, NippyJarError> {
+        let file = File::open(config.data_path())?;
+        let mmap = unsafe { Mmap::map(&file)? };
+        Ok(NippyJarCursor {
+            jar: config,
+            zstd_decompressors,
+            file_handle: file,
+            mmap_handle: mmap,
+            tmp_buf: vec![],
+            row: 0,
+            col: 0,
+        })
+    }
+
+    /// Resets cursor to the beginning.
+    pub fn reset(&mut self) {
+        self.row = 0;
+        self.col = 0;
+    }
+
+    /// Returns a row, searching it by an entry used during [`NippyJar::prepare_index`].
+    ///
+    /// **May return false positives.**
+    ///
+    /// Example usage would be querying a transactions file with a transaction hash which is **NOT**
+    /// stored in file.
+    pub fn row_by_filter(&mut self, value: &[u8]) -> Result<Option<Vec<Vec<u8>>>, NippyJarError> {
```

Comment:
nit: return type might be a bit hard to grasp because of nested vecs, may be worth introducing an alias like `type Value = Vec<u8>` and mentioning in the comments that it returns columns